PKL Config Scala

andyglow opened the pull request:
## Purpose

Adds `pkl-config-scala` — the binding/runtime layer that lets hand-authored Scala case classes consume Pkl configs. Mirrors `pkl-config-kotlin`. A follow-up will add `pkl-codegen-scala` so users can either hand-write or generate their Scala types.

```scala
import org.pkl.config.java.ConfigEvaluator
import org.pkl.config.scala.syntax._

case class AppConfig(name: String, port: Int, tags: Seq[String])

val config: AppConfig = ConfigEvaluator
  .preconfigured()
  .forScala()
  .evaluate(ModuleSource.text("""
    |name = "orders"
    |port = 8080
    |tags = List("beta", "team-a")
    |""".stripMargin))
  .to[AppConfig]
```

Built on Scala 3.3.7 (LTS). Consumable from Scala 2.13 callers via `-Ytasty-reader` — see [Scala 2.13 cross-support](#scala-213-cross-support) below.

## What's covered

Scala-specific mappings added by this module:

| Pkl type | Scala target |
|---|---|
| `class Foo { x: Int }` | `case class Foo(x: Int)` |
| `Duration` | `scala.concurrent.duration.Duration`, `FiniteDuration` |
| `Int` (epoch) / `String` (ISO-8601) | `java.time.Instant` |
| `String` / `Regex` | `scala.util.matching.Regex` |
| `Pair<A, B>` | `(A, B)` / `Tuple2[A, B]` |
| `List<T>`, `Set<T>`, `Listing<T>`, `Collection<T>` | `Seq[T]` |
| `Map<K, V>`, `Mapping<K, V>` | `immutable.Map[K, V]` |
| Nullable `T?` | `Option[T]` |
| String-literal union (Pkl codegen's enum form) | Scala 3 `enum Foo { case A, B }` — both plain and Java-compat `enum Foo extends java.lang.Enum[Foo]` forms |

Inherited transparently from `pkl-config-java` (regression-pinned in `ScalaObjectMapperSpec`):

- `Int` → `java.math.BigInteger`, `java.math.BigDecimal`, `Byte`, `Short`, `Int`, `Float`, `Double`
- `Float` → `Float`, `BigDecimal`
- `String` → `java.lang.Character`, `java.net.URI`, `java.net.URL`, `java.nio.file.Path`, `java.io.File`, `Pattern`
- `Bytes` → `Array[Byte]`
- `DataSize`, `Duration`, `Regex` identity passthroughs
- Java `enum` via `pStringToEnum`

### Intentionally not covered

Users with hand-authored types register their own converter factories:

- Scala 2 `scala.Enumeration` — out of scope. The Scala 2.13 reflection plumbing it required is what motivated removing the earlier enum path; with Scala 3 enums supported natively, the Scala 2 form is dead weight.
- Scala `Set[T]`, `Vector[T]`, mutable collections — matches the minimal default set in `pkl-config-java` / `pkl-config-kotlin`.
- Scala primitive `Char` — Java's `pStringToCharacter` registers for `java.lang.Character` only; case classes with a `Char` field need to use `java.lang.Character` instead, or register a `(String, char.class)` conversion.

## Module shape

Mirrors `pkl-config-kotlin`:

- `ScalaConverterFactories.pObjectToCaseClass` is `new PObjectToDataObject { override selectConstructor }` — same anonymous-subclass pattern as `KotlinConverterFactories.pObjectToDataObject`.
- `ScalaConversions` carries scalar conversions (Instant / Duration / Regex), analogous to `KotlinConversions`.
- `forScala()` extensions on `ValueMapperBuilder` / `ConfigEvaluatorBuilder` / `ConfigEvaluator` / `ConfigDecoderBuilder` / `ConfigDecoder`, plus `Config.to[T]`. Matches the five-receiver surface of `forKotlin()` in `ConfigExtensions.kt`.
- Parametric collection converters (`Option`, `Pair`, `Map`, `Seq`) go through a small `CachedConverterFactories` helper for per-type-arg caching. Same idea as Kotlin's `PPairToKotlinPair`, factored to keep the four factories declarative.
- `PStringToScalaEnum` handles idiomatic `enum Foo { case A, B }` by reading the companion module's synthetic `values()` method into an order-preserving `VectorMap`. For the opt-in `enum Foo extends java.lang.Enum[Foo]` form, `Class.isEnum` returns true and `pkl-config-java`'s stock `PStringToEnum` handles it; our factory defers in that case to avoid double-handling.

No `scala-reflect` dependency.

## Scala 2.13 cross-support

Built for Scala 3.3.7. Consumable from Scala 2.13 callers via the TASTy reader:

```scala
scalacOptions += "-Ytasty-reader"
libraryDependencies += "org.pkl-lang" %% "pkl-config-scala" % "X.Y.Z"
```

Works for the 2.13 consumer:
- `forScala()` builder extensions and `Config.to[T]` — kept as `implicit class` (pre-Scala 3 form), TASTy-readable.
- 2.13 case classes — implement `scala.Product`, which is what `pObjectToCaseClass` checks for.
- All collection / `Option` / `Pair` / regex / duration converters — use `scala.collection.immutable.*`, common to both Scala versions.

Does not work:
- Scala 2 `Enumeration` — never supported by this module.
- Scala 3 `enum` — 2.13 has no `enum` keyword; `PStringToScalaEnum` is dead-but-harmless code on the consumer side.

A follow-up could wire a smoke-test module that compiles against Scala 2.13 with `-Ytasty-reader` and consumes `pkl-config-scala_3`, to catch TASTy-reader regressions automatically.

## Notes

- Default converter set trimmed to match `pkl-config-java` / `pkl-config-kotlin`. Users register their own for `Set` / `Vector` / mutable collections.
- Scala-native types preferred internally (`TrieMap`, Scala `Option` + `.toJava` at Java boundaries). Java types kept only where the API demands them (`java.lang.reflect.*`, `MethodHandle`, `Optional` return).
- POM publishing wiring (the `library` vs `fatJar` ambiguity that `pkl-config-kotlin.gradle.kts` works around with manual XML injection) will be addressed in a follow-up; not required for this PR's scope.

bioball commented on buildSrc/src/main/kotlin/pklJavaLibrary.gradle.kts:
```diff
@@ -59,6 +59,14 @@ spotless {
+      rootProject.file("buildSrc/src/main/resources/license-header.star-block.txt"),
+      "package ",
+    )
+  }
```
Move this to `pklScalaLibrary.gradle.kts`

andyglow replied:
Done

bioball commented on build-logic/src/main/kotlin/pklScalaLibrary.gradle.kts:
```diff
@@ -0,0 +1,50 @@
+
+// Version Catalog library symbols.
+val libs = the<LibrariesForLibs>()
+
```
Add:

```kt
scala {
  scalaVersion = libs.versions.scala
}
```

andyglow replied:
Added `scala { scalaVersion = libs.versions.scala }`. Dep report confirms `scala-library:2.13.17` now lands on the compile classpath via the plugin's auto-managed dependency.

bioball commented on build-logic/src/main/kotlin/pklScalaLibrary.gradle.kts:
```diff
@@ -0,0 +1,50 @@
+val libs = the<LibrariesForLibs>()
+
+dependencies {
+  api(libs.scalaLibrary)
```
Remove (this is redundant when setting `scala.scalaVersion`)

```suggestion
```

andyglow replied:
Removed `api(libs.scalaLibrary)` — redundant now that the plugin wires it in.

bioball commented on pkl-config-scala/src/main/scala/org/pkl/config/scala/mapper/CachedSourceTypeInfo.scala:
```diff
@@ -0,0 +1,78 @@
+    PClassInfo.Unavailable.asInstanceOf[PClassInfo[Any]]
+
+  // Holds an optional converter, cached upon first retrieval.
+  private var converter: Option[Converter[Any, Any]] = None
```
These should be thread-safe, because `ValueMapper`s can possibly be shared across threads.

Maybe put inside `AtomicReference`?

andyglow replied:
Done. Merged the two `var`s into a single `AtomicReference[Entry(classInfo, converter)]` so the paired state updates coherently — no risk of a reader seeing a new `classInfo` with a stale `converter`. Under contention, threads may each compute a fresh converter (wasted `getConverter` lookup), but never see a torn state.

bioball commented on pkl-config-scala/src/main/scala/org/pkl/config/scala/mapper/PStringOrIntToEnumeration.scala:
```diff
@@ -0,0 +1,59 @@
+            case _ => None
+          }
+
+          res.orNull
```
Should throw `ConversionException` if no target types match (see `org.pkl.config.java.mapper.PStringToEnum`)

Also, I don't know if we should accept int to enum; in Java (and in every other binding), we only convert unions of string literals to enums.

bioball commented on pkl-config-scala/src/main/scala/org/pkl/config/scala/mapper/PStringOrIntToEnumeration.scala:
```diff
@@ -0,0 +1,59 @@
+  override def create(
+      sourceType: PClassInfo[_],
+      targetType: Type
+  ): Optional[Converter[_, _]] = {
```
Two comments:

1. This should return `Optional.empty` if `sourceType` is not `PClassInfo.String`
2. The return type should be `Optional[Converter[String, Enumeration]]` right? Is there any case where the mapped value wouldn't be an `Enumeration`?

andyglow replied:
Both points addressed, but the final shape diverged from the initial refactor. Enum params no longer go through the generic `ConverterFactory.create` path at all — they're dispatched directly via pattern match on `Param.Type.ScalaEnum` inside `ScalaPObjectToCaseClass` (see #12). So the question of whether to tighten the `Optional[Converter[_, _]]` signature is moot: the wildcard never participates in enum resolution anymore. The inner lookup is strongly typed: `String => Enumeration#Value`.

bioball commented on pkl-config-scala/src/main/scala/org/pkl/config/scala/mapper/PStringOrIntToEnumeration.scala:
```diff
@@ -0,0 +1,59 @@
+import java.util.Optional
+import scala.jdk.OptionConverters._
+
+private[mapper] object PStringOrIntToEnumeration extends ConverterFactory {
```
We shouldn't be mapping int types to enums (Pkl only turns union types of string literals into enums during codegen). So, let's call this `PStringToEnumeration`

andyglow replied:
Renamed, and the int branch is gone. Heads up: the whole enum-resolution path was reworked end-to-end as part of #12 — it's now driven by `ConstructorParamResolver` + a sealed `Param.Type` hierarchy, and the string→member lookup lives in a small `matchEnumMember` helper rather than a dedicated class.

bioball commented on build-logic/src/main/kotlin/pklAllProjects.gradle.kts:
```diff
@@ -26,11 +26,24 @@ dependencyLocking { lockAllConfigurations() }
 
 configurations {
   val rejectedVersionSuffix = Regex("-alpha|-beta|-eap|-m|-rc|-snapshot", RegexOption.IGNORE_CASE)
+  val versionSuffixRejectionExcemptions =
```
```suggestion
  val versionSuffixRejectionExemptions =
```

andyglow replied:
Fixed — applied to both the declaration and the usage site.

bioball commented on build-logic/src/main/kotlin/pklAllProjects.gradle.kts:
```diff
@@ -26,11 +26,24 @@ dependencyLocking { lockAllConfigurations() }
+    setOf(
+      // I know.
+      // This looks odd.
+      // But yes, it's transitively required by one of the relese versions of `zinc`
```
```suggestion
      // But yes, it's transitively required by one of the release versions of `zinc`
```

andyglow replied:
Fixed.

bioball commented on pkl-config-scala/src/main/scala/org/pkl/config/scala/mapper/PStringOrIntToEnumeration.scala:
```diff
@@ -0,0 +1,59 @@
+                        .toEnumConstantName(n)
+                        .equals(name)
+                    } =>
+                  value
```
This should throw `ConversionException` if the string doesn't match any of the member names

andyglow replied:
Done. `matchEnumMember` throws `ConversionException` with the input value plus the full candidate list when no member matches. Covered by the new `unknown enum value raises ConversionException listing valid members` test from #15.

bioball commented on pkl-config-scala/src/main/scala/org/pkl/config/scala/mapper/PStringOrIntToEnumeration.scala:
```diff
@@ -0,0 +1,59 @@
+      targetType: Type
+  ): Optional[Converter[_, _]] = {
+    targetType.asCustomEnum.map { members =>
+      (new Converter[Any, Any] {
```
Can you move this into its own class? 

The body of `create()` should just be something like:

```scala
    if (sourceType != PClassInfo.String) return Optional.empty()
    targetType.asCustomEnum.map(new ConverterImpl(_)).toJava
```

andyglow replied:
Extracted initially to `StringToEnumerationConverter`. In the final shape (after #12 reworked enum dispatch), the class collapsed into a private `matchEnumMember(value, members)` helper function — it's stateless and we no longer route through the `Converter` interface for enums, so the class wrapper was leftover scaffolding.

bioball commented on pkl-config-scala/src/main/scala/org/pkl/config/scala/mapper/ScalaConverterFactories.scala:
```diff
@@ -0,0 +1,246 @@
+    pCollectionToMutableBuffer,
+    pCollectionToMutableSeq,
+    pObjectToCaseClass,
+    PStringOrIntToEnumeration
```
We don't offer most of these conversions for Java and Kotlin. To be consistent, I think only the converters that pair with codegen should be here.

I think that's just these ones:

* pPairToTuple
* pMapToImmutableMap
* pCollectionToSeq
* pObjectToCaseClass
* pStringToEnumeration
* pAnyToOption

Users can always add their own converters if they are hand-authoring their own classes and need them.

bioball replied:
Also: not sure how we can do this, but it'd also be good to support strings to Scala 3 enums.

andyglow replied:
Trimmed. `all` is now:

```
pAnyToOption, pPairToTuple, pMapToImmutableMap, pCollectionToImmutableSeq, pObjectToCaseClass
```

Note: `pStringToEnumeration` isn't a separate factory anymore — enum dispatch is handled inside `ScalaPObjectToCaseClass` (see #12). The 13 other factory vals + two private helpers were removed. Test cases for Set/List/Vector/Stream/LazyList/all-mutable/PObject-as-Map were dropped accordingly.

andyglow replied:
Deferred to a follow-up. Scala 3 `enum` compiles to a `java.lang.Enum` subclass with `.values()` / `.valueOf(String)`, a different shape from Scala 2 `Enumeration`. The `Param.Type` sealed hierarchy is set up to accept a new variant (e.g. `Param.Type.JavaEnum`) alongside the current `Jvm` / `ScalaEnum` when the time comes. Happy to file an issue tracking this.

bioball commented on pkl-config-scala/src/main/scala/org/pkl/config/scala/syntax/package.scala:
```diff
@@ -0,0 +1,137 @@
+    */
+  implicit class ValueMapperBuilderSyntaxExtension(val x: ValueMapperBuilder)
+      extends AnyVal {
+    def forScala(): ValueMapperBuilder = {
```
Nice!

bioball commented on pkl-config-scala/src/test/scala/org/pkl/config/scala/ScalaObjectMapperSpec.scala:
```diff
@@ -0,0 +1,279 @@
+class ScalaObjectMapperSpec extends AnyFunSuite {
+  import ScalaObjectMapperSpec._
+
+  test("evaluate scala types") {
```
Let's add some tests for failure cases; some to consider:

* Incorrectly mapped data class
* unknown enum values
* Type mismatch with `Config.to[T]`

andyglow replied:
Added tests for all three scenarios you listed:

- `unknown enum value raises ConversionException listing valid members` — asserts the bad input + full candidate list appear in the message
- `missing required property on case class raises ConversionException` — asserts the missing field name is in the message
- `type mismatch between Pkl value and case class field raises ConversionException` — asserts a "cannot convert" hint

Plus one more — `enum property receiving non-String Pkl value raises ConversionException` — covering the defensive branch inside `ScalaPObjectToCaseClass`.

bioball commented on pkl-config-scala/src/main/scala/org/pkl/config/scala/mapper/ScalaConverterFactories.scala:
```diff
@@ -0,0 +1,246 @@
+  //    .forParametrizedType1[java.util.Collection[_], Array[_]](
+  //      x => x == PClassInfo.Collection | x == PClassInfo.Set | x == PClassInfo.List,
+  //      pCollectionToMutableCollectionConv[Array](_.iterator.toArray[Any])
+  //    )
```
Remove unused code

andyglow replied:
Removed — both the commented-out block and its entry in `all`.

bioball commented on pkl-config-scala/NOTE.md:
```diff
@@ -0,0 +1,44 @@
+- `Either` (???)
+- `sealed traits`
+- `object` instances
+- cross-version compilation to cover scala `2.12` too
```
Accidentally committed this file?

andyglow replied:
Yes — accidentally committed. Deleted.

bioball commented on pkl-config-scala/src/main/scala/org/pkl/config/scala/mapper/ScalaConverterFactories.scala:
```diff
@@ -0,0 +1,246 @@
+
+  val pObjectToImmutableMap: ConverterFactory =
+    CachedConverterFactories.forParametrizedType2[PObject, immutable.Map[_, _]](
+      x => x == PClassInfo.Object | x == PClassInfo.Dynamic,
```
Here, and everywhere else: use logical OR (`||`), not bitwise OR, because:

1. Easier to read and understand
2. Short-circuiting

andyglow replied:
Swept all 13 occurrences in this file. Left untouched the pattern-match alternative (`case _: PNull | null`) and `|` in multi-line-string margins — those aren't boolean ORs.

bioball commented on pkl-config-scala/src/main/scala/org/pkl/config/scala/mapper/JavaReflectionSyntaxExtensions.scala:
```diff
@@ -0,0 +1,134 @@
+      *   Some(list of `Enumeration#Value`) if the enumeration can be resolved,
+      *   None otherwise
+      */
+    def asCustomEnum: Option[List[Enumeration#Value]] = {
```
I'm not totally following the code here.

Why do we need to care about case classes here? Can we expect enums to simply look like this instead?

```scala
object SimpleEnum extends Enumeration {
  val Aaa, Bbb, Ccc = Value
}
```

And we can just use `.values` to access enum members during conversion.

Pkl only turns string literal unions into enums in every other language; this means that we don't need our enum values to be parameterized.

andyglow replied:
Done — and the annotation is deleted, not just optional. Both forms work now:

```scala
// plain — previously unsupported
object Color extends Enumeration { val Red, Green, Blue = Value }
case class Palette(primary: Color.Value)

// nested-Val form, still supported, no annotation needed
object Shape extends Enumeration {
  case class V() extends Val(nextId)
  val Circle, Square = V()
}
case class Drawing(shape: Shape.V)
```

Getting there required more than `.getDeclaringClass`. For the plain form the runtime class is `scala.Enumeration$Value`, shared across every `Enumeration` in the JVM — Java reflection can't distinguish owners. I explored three workarounds before settling on the current approach:

1. Pre-substitute resolved `Value` instances into the `PObject`'s property map and delegate. Dead — `PClassInfo.forValue` throws on non-Pkl values.
2. Swap enum-param `Type`s for a sentinel `ParameterizedType` carrying the `Enumeration` reference. Dead — `ValueMapperImpl.getConverter` calls `Reflection.getExactSubtype(...)` before factory dispatch, which normalises custom `Type`s back to their raw class.
3. Upstream patch (make `create` non-final / add a per-param hook). Out of scope for this PR.

What we landed on:

- `ConstructorParamResolver` uses `scala.reflect.runtime.universe` only to walk ctor params and recover the path-dependent `Enumeration#Value` prefix. It emits pure-data `Param`s with a sealed `Param.Type { Jvm | ScalaEnum }`.
- `ScalaPObjectToCaseClass` forks `PObjectToDataObject.ConverterImpl` and dispatches on `Param.Type` directly. For `ScalaEnum` it bypasses `ValueMapper.getConverter` (which can't see past the erasure) and does a direct member lookup.

The fork is the main cost — roughly 150 lines duplicated from `PObjectToDataObject.ConverterImpl.convert`. Happy to collapse it if you'd accept an upstream patch exposing a per-param hook; that would reduce this module to a slim subclass. Architectural details are in the updated PR description.

bioball reviewed (changes requested):
Made a first pass. Thanks for the contribution, this looks like a great start!

bioball commented on pkl-config-scala/src/main/scala/org/pkl/config/scala/mapper/CachedConverterFactories.scala:
```diff
@@ -0,0 +1,163 @@
+
+/** Provides infrastructure that helps define custom converter factories in a
+  * somewhat concise way at the same time utilizing caching.
+  */
```
Here, and everywhere else: can we use javadoc style doc comments? E.g.

```scala
/**
 * Provides infrastructure that helps define custom converter factories in a
 * somewhat concise way at the same time utilizing caching.
 */
```

I think this means setting up a `.scalafmt.conf` file.

andyglow replied:
Done. Added `.scalafmt.conf` at repo root with `docstrings.style = Asterisk`, `docstrings.blankFirstLine = yes`, `docstrings.wrap = yes`, and wired spotless to use it via `scalafmt(version).configFile(...)`. `spotlessApply` reformatted every docstring in the module to match the example in your comment.

Commit:
PKL Config Scala

ktham commented:
Really excited to see some momentum on Scala support for pkl! I think it could be a good alternative to [lightbend/config](https://github.com/lightbend/config) and [pureconfig](https://github.com/pureconfig/pureconfig) if there were first-class Scala support.

I see this PR specifically targets Scala 2.13. One thing that stands out to me is that it isn't clear what would happen to code that won't compile under Scala 3, e.g. `scala.reflect.runtime.universe`. Would we have separate directories for Scala 2-specific code? And what would the Gradle build look like?

Given that Scala 3 was released ~5 years ago and most actively maintained Scala projects now cross-compile, it might be worth attempting a Scala 3 cross-build before merging the initial Scala 2 support to surface any surprises early and to have clarity on whether there are any known blockers.

Either way, if Scala support is being prioritized for an upcoming pkl release, I'd love to see a Scala 3 variant simultaneously land as part of that same release 🙂 (I'd be happy to take a look and help with that!)

andyglow commented:
Hello @ktham.
The plans are wide. We will start from a reflection based (to mimic approaches taken for java and kotlin) minimal implementation for `2.13` (+ `scala 3` cross-compilation). 

The plan is though to than gradually drift to macro based solution. @bioball correct me if i'm wrong.
I also have a pureconfig-pkl prototype that we potentially can add later on. And maybe for Ciris. 
As well as there is a plan for adding scalameta based codegen.

bioball commented:
On `scala.reflect.runtime.universe`: the reason this is used is to support decoding `Enumeration` objects. However, we've been chatting about how to best support enums in Scala, and I'm kind of bearish on supporting `Enumeration`; our object mapper currently relies on Java reflection, and Scala's `Enumeration` values are erased into just `scala.Enumeration$Value`.

And yeah; I think we do want to move towards generated deserializers for JVM languages. I think this would be implemented as annotation processors on the Java/Kotlin side. Maybe macros for Scala, although I think that might also be challenging if we want to support both Scala 2 and 3.

odenix commented:
For a library first released in 2026, I’d focus on Scala 3 and wouldn’t support Scala 2 anymore. I’m also worried that adding pkl-scala to this repo could make the Gradle build much heavier, like the internal IntelliJ plugin did before being moved out. (I’d also love to see pkl-gradle moved out, which currently dominates build time.)

andyglow commented:
hi folks. thanks for the feedbacks. appreciate it

following your recommendations i have rewritten it to scala 3
so scala 3 is a first class citizen. scala 3 enums got supported too.

with `-Ytasty-reader` its enabling scala 2.13 projects too (no enum support for scala 2 though)

bioball commented on pkl-config-scala/src/main/scala/org/pkl/config/scala/syntax/package.scala:
```diff
@@ -0,0 +1,123 @@
+   *   }}}
+   */
+  implicit class ConfigSyntaxExtension(val x: Config) extends AnyVal {
+    def to[T](implicit ct: ClassTag[T]): T = x.as[T](ct.runtimeClass)
```
This will lose any type arguments to `T`; e.g. `to[Seq[Int]]`.

I think there's a couple ways around this:

1. We copy what we do in Java (see org.pkl.config.java.JavaType), and offer a `ScalaType` API
2. We implement this as a macro (for both scala 2 and 3 if we want to ensure continued 2.13 support)
3. We use izumi-reflect's `Tag`

bioball replied:
Ah, this also doesn't handle union types, or string literal types, e.g.

```scala
config.to["hello"] // downgrades to `to[String]`
config.to[Int | Boolean] // downgrades to `to[Object]`
```

andyglow replied:
yes.. definitely. i missed that as we don't have cases for it. but it is definitely a gap
the easiest fix would be to incorporate `izumi`. but.. are we ok for external dependency?
we definitely can solve on our own. it will require some boilerplate though if we want to support 2.13 as the fix that will work for scala 3 will not be accepted by 2.13 even if it's a `3.7` sadly. so we will need cross-version compilation with all the corresponding "benefits"

let me create tests and solve it for just 3 at this point and we will see what to do next

bioball commented on pkl-config-scala/src/main/scala/org/pkl/config/scala/mapper/ScalaConversions.scala:
```diff
@@ -0,0 +1,85 @@
+/**
+ * Provides conversions between Java types backing PKL and Scala types, enabling seamless
+ * interoperability for configuration values within PKL.
+ */
```
```suggestion
 */
 //noinspection ScalaWeakerAccess
```

bioball commented on pkl-config-scala/src/main/scala/org/pkl/config/scala/mapper/ScalaConversions.scala:
```diff
@@ -0,0 +1,85 @@
+    Conversion.of(
+      PClassInfo.Duration,
+      classOf[Duration],
+      (v: PDuration, _) => Duration.fromNanos(v.inNanos()).toCoarsest
```
```suggestion
      (v: PDuration, _) => {
        val unit = v.getUnit match {
          case DurationUnit.NANOS => TimeUnit.NANOSECONDS
          case DurationUnit.MICROS => TimeUnit.MICROSECONDS
          case DurationUnit.MILLIS => TimeUnit.MILLISECONDS
          case DurationUnit.SECONDS => TimeUnit.SECONDS
          case DurationUnit.MINUTES => TimeUnit.MINUTES
          case DurationUnit.HOURS => TimeUnit.HOURS
          case DurationUnit.DAYS => TimeUnit.DAYS
        }
        Duration(v.getValue, unit)
      }
```

andyglow replied:
mmm. ok. explicit it is

bioball commented on pkl-config-scala/src/main/scala/org/pkl/config/scala/mapper/ScalaConverterFactories.scala:
```diff
@@ -0,0 +1,120 @@
+            case _: PNull | null => None
+            case v: Option[_]    => v.map(s1.updateAndGet(_, t1, vm))
+            case v: Optional[_]  => v.toScala.map(s1.updateAndGet(_, t1, vm))
+            case v               => Option(s1.updateAndGet(v, t1, vm))
```
You won't get either `Option` or `Optional` from Pkl.

Also: `Some` constructor feels more idiomatic

```suggestion
            case v               => Some(s1.updateAndGet(v, t1, vm))
```

bioball commented on pkl-config-scala/src/main/scala/org/pkl/config/scala/mapper/ScalaConverterFactories.scala:
```diff
@@ -0,0 +1,120 @@
+          clazz.getInterfaces
+            .exists(i => classOf[scala.Product].isAssignableFrom(i))
+        })
+        .toJava
```
Better to delegate to Java implementation (find constructor whose argument count matches field count) in case Scala synthesizes multiple constructors per case class.

```suggestion
      // case classes all implement Product
      if (!clazz.getInterfaces.exists(i => classOf[scala.Product].isAssignableFrom(i))) {
        return Optional.empty()
      }
      super.selectConstructor(clazz)
```

bioball commented on pkl-config-scala/src/test/scala/org/pkl/config/scala/mapper/PPairToScalaTupleSpec.scala:
```diff
@@ -0,0 +1,106 @@
+import scala.jdk.CollectionConverters._
+
+class PPairToScalaTupleSpec extends AnyFunSuite with BeforeAndAfterAll {
+  import PPairToScalaTupleSpec._
```
There's a couple more places in the code that uses `._` style wildcard imports

```suggestion
import org.scalatest.matchers.should.Matchers.*
import org.pkl.config.scala.syntax.*

import scala.jdk.CollectionConverters.*

class PPairToScalaTupleSpec extends AnyFunSuite with BeforeAndAfterAll {
  import PPairToScalaTupleSpec.*
```

bioball commented on pkl-config-scala/src/test/scala/org/pkl/config/scala/mapper/PPairToScalaTupleSpec.scala:
```diff
@@ -0,0 +1,106 @@
+      mapper.map(
+        ex1,
+        Types.parameterizedType(
+          classOf[Tuple2[_, _]],
```
Same here; some more places that use `_` that should be replaced with `?`

```suggestion
          classOf[Tuple2[?, ?]],
```

bioball commented on pkl-config-scala/src/main/scala/org/pkl/config/scala/mapper/PStringToScalaEnum.scala:
```diff
@@ -0,0 +1,71 @@
+    values.iterator.collect { case v: scala.reflect.Enum =>
+      v.productPrefix -> v.asInstanceOf[AnyRef]
+    }.toMap
+  }.toOption
```
Why wrap this in `Try {}.toOption`? What do we anticipate to fail here? Can we catch more specific errors?

andyglow replied:
hmm. good point. let's rethrow than. probably cleaner
all the masked errors are reflective here

bioball commented on pkl-config-scala/src/main/scala/org/pkl/config/scala/mapper/PStringToScalaEnum.scala:
```diff
@@ -0,0 +1,71 @@
+ *
+ * Scala 3 enum cases implement `scala.reflect.Enum`. When the enum also extends `java.lang.Enum`
+ * (the opt-in Java-compat form), `Class.isEnum` returns true and `pkl-config-java`'s
+ * `PStringToEnum` already handles them; this factory defers in that case. For the plain
```
I don't see any code here that defers?

bioball commented on pkl-config-scala/src/main/scala/org/pkl/config/scala/mapper/ScalaConverterFactories.scala:
```diff
@@ -0,0 +1,120 @@
+ * hand-author Scala classes with other collection shapes (Set, Vector, mutable collections, etc.)
+ * can register their own converter factories.
+ */
+object ScalaConverterFactories {
```
```suggestion
//noinspection ScalaWeakerAccess
object ScalaConverterFactories {
```

bioball commented on pkl-config-scala/src/test/scala/org/pkl/config/scala/ScalaObjectMapperSpec.scala:
```diff
@@ -0,0 +1,289 @@
+      assert(result == Person("via-builder", 7))
+    } finally {
+      evaluator.close()
+    }
```
[nit] share the same evaluator across all tests

If we can't, then do `Using(ConfigEvaluatorBuilder.preconfigured().forScala().build()) { evaluator =>`

bioball commented on pkl-config-scala/pkl-config-scala.gradle.kts:
```diff
@@ -0,0 +1,33 @@
+    named<MavenPublication>("library") {
+      pom {
+        url.set("https://github.com/apple/pkl/tree/main/pkl-config-scala")
+        description.set("Scala config library based on the Pkl config language.")
```
```suggestion
        description.set("Scala extensions for pkl-config-java, a Java config library based on the Pkl config language.")
```

andyglow replied:
this is a nice point to discuss
if you remember well my very first version of this lib was built using scala macros (2.13 though)
it didn't require pkl-config-java. than there was ask to unify and now we're closing the loop. which is fine, but i want to emphasize that bringing two toolsets for solving one problem is even more confusing.
per my understanding it's either
- pkl-config-java extension and then it uses Conversions/ConversionFactories and the stuff all over the stack
- or it's pure scala with no reflection at all - macro/shapeless/magnolia/izumi - this kind of stack, no java reflection no dependency on pkl-config-java

what do you think?

bioball commented on build-logic/src/main/kotlin/pklScalaLibrary.gradle.kts:
```diff
@@ -0,0 +1,60 @@
+    includeEngines("scalatest")
+    testLogging { events("passed", "skipped", "failed") }
+  }
+}
```
Not necessarily blocking this PR: if our goal is to support both Scala 2.13 and Scala 3, we should be testing the library from both targets.

andyglow replied:
i think we should produce scala 3.. 
scala 2.13 consumers can pull scala 3 library
https://www.scala-lang.org/blog/state-of-tasty-reader.html

so at this point we can lock at `3.7` and scale ro `2.13` and `3.8` later

bioball reviewed (changes requested):
Did another pass! This is looking great! I think the biggest problem here is that `ClassTag` doesn't give us type arguments of generic types.

bioball commented on build-logic/src/main/kotlin/pklScalaLibrary.gradle.kts:
```diff
@@ -0,0 +1,60 @@
+ * limitations under the License.
+ */
+@file:Suppress("HttpUrlsUsage", "unused")
+
```
```suggestion
```

Commit:
Remove Scala 2 Enumeration support

Drops the scala-reflect dependency and the bespoke
ConstructorParamResolver / Param / ScalaPObjectToCaseClass plumbing that
existed only to recover path-dependent Enumeration#Value types past Java
reflection erasure. Reverts pObjectToCaseClass to the inline
PObjectToDataObject anonymous subclass shape used by pkl-config-kotlin.

Phase-1 scope; Scala 3 enum support follows in a later commit on this branch.

Commit:
Migrate to Scala 3.3.7; add Scala 3 enum support

Build:
- Bump scala = "3.3.7" (LTS); switch scala-library, scalatest,
  scalatestplus-junit, diffx artifacts to their _3 variants
- Drop -Xsource:3 (Scala 2 dialect flag); -release:N alone enforces
  bytecode level on the Scala 3 compiler

Source: zero changes from Scala 2.13 source — the prior -Xsource:3
discipline already had the module on the Scala 3 dialect.

Adds PStringToScalaEnum factory for idiomatic `enum Foo { case A, B }`,
which Class.isEnum doesn't recognize (Scala 3 only sets ACC_ENUM when
the user opts in via `extends java.lang.Enum`). Looks up cases via the
companion module's synthetic values() method. Defers when Class.isEnum
is true so pkl-config-java's stock PStringToEnum handles the
Java-compat form, avoiding double-handling.

No scala-reflect dependency.

Commit:
Polish: ConfigDecoder extensions, regression-pin tests, order-preserving enum lookup

- Add forScala() extensions on ConfigDecoderBuilder and ConfigDecoder,
  matching pkl-config-kotlin's five-receiver surface.
- Switch PStringToScalaEnum's internal store from Map to VectorMap so
  the error message lists candidate cases in enum-declaration order.
- Extract dedicated `idiomatic Scala 3 enum routes through PStringToScalaEnum`
  test (was previously buried in the kitchen-sink test).
- Strengthen the unknown-enum-value test to assert declaration order.
- Add regression-pin test covering inherited Java conversions:
  BigInteger, BigDecimal, URI, URL, Path, File, Character, Bytes,
  DataSize, Listing, Mapping. Catches accidental factory-ordering
  regressions and documents that primitive Char isn't covered (only
  java.lang.Character is).

Commit:
Polish: drop dead @file:Suppress, add ScalaWeakerAccess on public-by-design objects

ScalaConversions and ScalaConverterFactories expose val members that are only
referenced internally from .all but are public-by-design (users may register
a subset). IntelliJ flags these as ScalaWeakerAccess; suppress to silence the
noise. The pklScalaLibrary @file:Suppress("HttpUrlsUsage", "unused") was
carried over from a template — file has no HTTP URLs and no unused declarations.

Commit:
Polish: replace `_` wildcards with `*`/`?` in pkl-config-scala tests

Main module already uses Scala 3 idiom (`*` for import wildcards, `?` for
type wildcards); the test files lagged. Sweep imports and `Tuple2[_, _]`.

Commit:
Polish: clarify PStringToScalaEnum scaladoc; align POM description

The "this factory defers" wording read like a method call; rewrite to spell
out the actual mechanism (return Optional.empty so the chain falls through).
POM description now mirrors pkl-config-kotlin's, framing the artifact as
extensions to pkl-config-java rather than a standalone library.

Commit:
Drop dead Option/Optional branches in pAnyToOption; use Some

Pkl never emits Option or java.util.Optional values — nullability surfaces as
PNull. The two reflective branches were copy-pasted defensively and could
never fire. The remaining catch-all is reachable only after PNull/null is
ruled out, so Option(...) was redundant null-checking; Some(...) makes the
intent explicit.

Commit:
Delegate case-class constructor selection to PObjectToDataObject.super

getDeclaredConstructors.headOption picked an arbitrary constructor when a
case class declares secondary constructors (user-written or synthesized).
Filter on the Product interface, then defer to the Java base class's
selectConstructor, which picks by argument count and matches the existing
pkl-config-java policy. Drops the now-unused OptionConverters import.

Commit:
Preserve PDuration unit in pDurationToDuration / pDurationToFiniteDuration

Both converters previously collapsed the value to nanos and let toCoarsest
reverse-engineer a unit by checking divisibility. Use PDuration's own unit
when the value fits in a Long (the common case for Pkl-emitted durations);
fall back to nanos+toCoarsest only for fractional values. As a side effect,
durations like `5000000.ns` now round-trip as NANOSECONDS instead of being
re-coarsified to MILLISECONDS — equivalent length, faithful to the source.

Commit:
Revert POM description change for pkl-config-scala

Reviewer's suggested wording ("Scala extensions for pkl-config-java...")
was a partial framing — the artifact does delegate to pkl-config-java but
also exposes Scala-native API surface (case-class targeting, syntax
package, Scala 3 enum support). Keeping the original.

Commit:
Narrow Try in PStringToScalaEnum.readEnumValues; rethrow with context

The Try { ... }.toOption pattern caught Throwable (including OOMEs and
StackOverflows) and swallowed the cause, leaving "no converter found"
upstream as the only signal of a misconfigured classpath. The block can
only realistically throw four things — ClassNotFoundException,
NoSuchFieldException, NoSuchMethodException, IllegalAccessException — all
ReflectiveOperationException. Catch that, then rethrow as
ConversionException with the failing class name and the original cause.

Commit:
Migrate ScalaObjectMapperSpec and PPairToScalaTupleSpec to FixtureAnyFunSuite

Both specs now build their evaluator (and dependent fixtures) inside
withFixture per test, with deterministic close in finally. ScalaObjectMapperSpec's
fixture is the ConfigEvaluator; PPairToScalaTupleSpec's bundles the evaluated
PModule and the configured ValueMapper since every test consumes both.

The ConfigEvaluatorBuilder-extension test stays separate — its purpose is to
exercise the builder path, not the evaluator one — and now uses scala.util.Using
for the same try-with-resources discipline as withFixture.

Commit:
spotlessApply: reflow doc, drop stray blank line
